src/data/matrix/basic.lean

```diff
@@ -26,24 +26,120 @@ open dmatrix
+  is_scalar_tower R S (matrix m n α) := pi.is_scalar_tower
+instance [monoid R] [mul_action R α] :
+  mul_action R (matrix m n α) := pi.mul_action _
+instance [monoid R] [add_monoid α] [distrib_mul_action R α] :
+  distrib_mul_action R (matrix m n α) := pi.distrib_mul_action _
+instance [semiring R] [add_comm_monoid α] [module R α] :
+  module R (matrix m n α) := pi.module _ _ _
+
 section ext
+
 variables {M N : matrix m n α}
 
 theorem ext_iff : (∀ i j, M i j = N i j) ↔ M = N :=
 ⟨λ h, funext $ λ i, funext $ h i, λ h, by simp [h]⟩
 
-@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N :=
-ext_iff.mp
+@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N := ext_iff.mp
 
 end ext
 
+section transpose
+
+/-- The transpose of a matrix. -/
+def transpose (M : matrix m n α) : matrix n m α
+| x y := M y x
+localized "postfix `ᵀ`:1500 := matrix.transpose" in matrix
+
+lemma trans_col_eq_row (A : matrix m n α) (i : m) : (λ j, Aᵀ j i) = A i :=
+by simp [transpose]
+lemma trans_row_eq_col (A : matrix m n α) (j : n) : Aᵀ j = (λ i, A i j):=
+by ext; simp [transpose]
+
+instance vec_has_star [has_star α]: has_star (n → α) := ⟨λ v i, star (v i)⟩
+@[simp] lemma vec_star_ext [has_star α] (v : n → α) (i : n) : star v i = star (v i) := rfl
+
+def conj_transpose [has_star α] (M : matrix m n α) : matrix n m α
+| x y := star (M y x)
+localized "postfix `ᴴ`:1500 := matrix.conj_transpose" in matrix
+
+end transpose
+
+open_locale matrix
+
+protected def is_sym (A : matrix m m α) : Prop := Aᵀ = A
+
+protected def is_skewsym [has_neg α] (A : matrix m m α) : Prop := -Aᵀ = A
+
+protected def is_Hermitian [has_star α] (A : matrix m m α) : Prop := Aᴴ = A
+
+/-- Proposition `matrix.is_perm`.
+    `P.is_perm` means `P` is a permutation matrix.
+    A new file is probably required for systematically investigating permutation matrices. -/
+protected def is_perm [decidable_eq m] [has_zero α] [has_one α] (P : matrix m m α): Prop :=
+∃ σ : equiv.perm m, P = equiv.perm.to_matrix σ
+
+/-- Proposition `matrix.is_perfect_shuffle`.
+    `P.is_perfect_shuffle` means `P` is a perfect shuffle matrix.
+    A new file is probably required for systematically investigating perfect shuffle matrices. -/
+protected def is_perfect_shuffle
+[decidable_eq m] [has_zero α] [has_one α] (P : matrix m m α) : Prop :=
+∃ σ : equiv.perm m, (P = equiv.perm.to_matrix σ ∧ ∀ i : m, σ i ≠ i)
+
+/-- The "trace" of a matrix.
+    A different version of "trace" is defined in `trace.lean` as `matrix.trace`.
+    One advantage of `matrix.tr` is that this definition is more elementary,
+    which only requires α to be a `add_comm_monoid`; whilst `matrix.trace` requires
+    `[semiring β] [add_comm_monoid α] [module β α]`.
+    The equivalence can be easily established when `α` is indeed a `β-module`.
+    Another advantage is that `matrix.tr` is more convenient for users to explore lemmas/theorems
+    involving "trace" from a combinatorial aspect.-/
+def tr [add_comm_monoid α] (A : matrix n n α) : α := ∑ i : n, A i i
```

Comment:
I don't buy this argument. 

> One advantage of `matrix.tr` is that this definition is more elementary, which only requires α to be a `add_comm_monoid`; whilst `matrix.trace` requires  `[semiring β] [add_comm_monoid α] [module β α]`.

Ever `add_comm_monoid α` satisfies the requirements for `matrix.trace` with `β = ℕ`.

> Another advantage is that `matrix.tr` is more convenient for users to explore lemmas/theorems

`matrix.trace_apply` contains exactly this statement of `trace

Reply:
Right, your argument makes sense. I did not notice `Ever add_comm_monoid α satisfies the requirements for matrix.trace with β = ℕ`, though `matrix.tr` is simpler in terms of it takes one less argument and has no coercion associated (in `matrix.trace_apply`, we need to use coercion).
Perhaps delete this simpler def?

Reply:
Yes, I think we should just delete your `tr`

> . I did not notice Ever add_comm_monoid α satisfies the requirements for matrix.trace with β = ℕ, though matrix.tr is simpler in terms of it takes one less argument

Perhaps the we should edit the docs for `trace` to remind users that they can use `β = ℕ`

> and has no coercion associated

This doesn't make things more complex in any meaningful way.

Reply:
OK. I will delete it.